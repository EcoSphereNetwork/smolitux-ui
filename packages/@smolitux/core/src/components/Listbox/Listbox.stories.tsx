import type { Meta, StoryObj } from '@storybook/react';
import { Listbox } from './Listbox';

const meta: Meta<typeof Listbox> = {
  title: 'Components/Core/Listbox',
  component: Listbox,
  tags: ['autodocs'],
  argTypes: {
    onChange: { action: 'change' },
    value: { control: 'text' },
    options: { control: 'object' },
    disabled: { control: 'boolean' },
    placeholder: { control: 'text' },
    autoFocus: { control: 'boolean' },
  },
  args: {
    value: '',
    options: [
      { label: 'Option 1', value: '1' },
      { label: 'Option 2', value: '2' },
      { label: 'Option 3', value: '3' },
    ],
  },
};
export default meta;

type Story = StoryObj<typeof Listbox>;

export const Default: Story = {};

export const WithPlaceholder: Story = {
  args: { placeholder: 'Select an option' },
};

export const Disabled: Story = {
  args: { disabled: true },
};

export const ManyOptions: Story = {
  args: {
    options: Array.from({ length: 20 }, (_, i) => ({
      label: `Option ${i + 1}`,
      value: String(i + 1),
    })),
  },
};

export const WithDisabledOption: Story = {
  args: {
    options: [
      { label: 'Active', value: 'a' },
      { label: 'Disabled', value: 'd', disabled: true },
    ],
  },
};
