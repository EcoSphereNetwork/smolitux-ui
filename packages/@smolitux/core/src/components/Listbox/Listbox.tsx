import React, { forwardRef, useEffect, useId, useState } from 'react';
import { cn } from '@smolitux/utils';

export interface ListboxOption {
  label: string;
  value: string;
  disabled?: boolean;
}

export interface ListboxProps {
  /** Selected value */
  value: string;
  /** Change handler */
  onChange: (value: string) => void;
  /** Available options */
  options: ListboxOption[];
  /** Disable entire listbox */
  disabled?: boolean;
  /** Accessible label */
  ariaLabel?: string;
  /** aria-describedby reference */
  ariaDescribedby?: string;
  /** Auto focus on mount */
  autoFocus?: boolean;
  /** id attribute */
  id?: string;
  /** Placeholder text if no value */
  placeholder?: string;
  /** Additional className */
  className?: string;
}

export const Listbox = forwardRef<HTMLDivElement, ListboxProps>(
  (
    {
      value,
      onChange,
      options,
      disabled = false,
      ariaLabel,
      ariaDescribedby,
      autoFocus = false,
      id,
      placeholder = 'No options',
      className,
      ...rest
    },
    ref,
  ) => {
    const uid = useId();
    const listId = id || `listbox-${uid}`;
    const [active, setActive] = useState(() =>
      Math.max(0, options.findIndex((o) => o.value === value)),
    );

    useEffect(() => {
      if (autoFocus && ref && 'current' in ref && (ref as any).current) {
        (ref as any).current.focus();
      }
    }, [autoFocus, ref]);

    useEffect(() => {
      const idx = options.findIndex((o) => o.value === value);
      if (idx !== -1) setActive(idx);
    }, [value, options]);

    function nextEnabled(start: number, dir: 1 | -1) {
      let i = start;
      while (i >= 0 && i < options.length) {
        if (!options[i].disabled) return i;
        i += dir;
      }
      return start;
    }

    function onKeyDown(e: React.KeyboardEvent<HTMLDivElement>) {
      if (disabled) return;
      if (e.key === 'ArrowDown') {
        e.preventDefault();
        setActive((i) => nextEnabled(Math.min(i + 1, options.length - 1), 1));
      } else if (e.key === 'ArrowUp') {
        e.preventDefault();
        setActive((i) => nextEnabled(Math.max(i - 1, 0), -1));
      } else if (e.key === 'Home') {
        e.preventDefault();
        setActive(nextEnabled(0, 1));
      } else if (e.key === 'End') {
        e.preventDefault();
        setActive(nextEnabled(options.length - 1, -1));
      } else if (e.key === 'Enter' || e.key === ' ') {
        e.preventDefault();
        const opt = options[active];
        if (opt && !opt.disabled) {
          onChange(opt.value);
        }
      }
    }

    const activeId = options[active] ? `${listId}-option-${active}` : undefined;

    return (
      <div
        ref={ref}
        id={listId}
        role="listbox"
        tabIndex={disabled ? -1 : 0}
        aria-activedescendant={activeId}
        aria-disabled={disabled || undefined}
        aria-label={ariaLabel}
        aria-describedby={ariaDescribedby}
        className={cn('smx-listbox border rounded', disabled && 'opacity-50', className)}
        onKeyDown={onKeyDown}
        {...rest}
      >
        {options.length === 0 ? (
          <div className="p-2 text-gray-500" role="option" aria-disabled="true">
            {placeholder}
          </div>
        ) : (
          options.map((opt, i) => {
            const selected = value === opt.value;
            return (
              <div
                key={opt.value}
                id={`${listId}-option-${i}`}
                role="option"
                aria-selected={selected}
                aria-disabled={opt.disabled || undefined}
                onMouseDown={(e) => e.preventDefault()}
                onClick={() => !opt.disabled && onChange(opt.value)}
                className={cn(
                  'px-3 py-1 cursor-pointer',
                  selected && 'bg-primary-500 text-white',
                  !selected && active === i && 'bg-primary-100',
                  opt.disabled && 'opacity-50 cursor-not-allowed',
                )}
              >
                {opt.label}
              </div>
            );
          })
        )}
      </div>
    );
  },
);

Listbox.displayName = 'Listbox';
export default Listbox;
