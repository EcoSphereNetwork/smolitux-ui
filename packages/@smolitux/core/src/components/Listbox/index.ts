export { Listbox as default } from './Listbox';
export type { ListboxProps, ListboxOption } from './Listbox';
