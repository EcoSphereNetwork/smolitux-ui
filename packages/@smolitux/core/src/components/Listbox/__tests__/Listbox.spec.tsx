import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';
import { Listbox } from '../Listbox';

const options = [
  { label: 'One', value: '1' },
  { label: 'Two', value: '2' },
  { label: 'Three', value: '3' },
];

describe('Listbox', () => {
  test('snapshot', () => {
    const tree = renderer
      .create(
        <Listbox value="1" onChange={() => {}} options={options} />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('selects option on click', async () => {
    const user = userEvent.setup();
    const handle = jest.fn();
    render(<Listbox value="1" onChange={handle} options={options} />);
    await user.click(screen.getByText('Two'));
    expect(handle).toHaveBeenCalledWith('2');
  });

  test('keyboard navigation', async () => {
    const user = userEvent.setup();
    const handle = jest.fn();
    render(
      <Listbox value="1" onChange={handle} options={options} aria-label="demo" />,
    );
    const list = screen.getByRole('listbox');
    list.focus();
    await user.keyboard('[ArrowDown]');
    await user.keyboard('[Enter]');
    expect(handle).toHaveBeenCalledWith('2');
  });
});
