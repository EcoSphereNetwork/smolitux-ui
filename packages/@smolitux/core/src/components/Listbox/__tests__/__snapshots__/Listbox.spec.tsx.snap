// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Listbox snapshot 1`] = `
<div
  aria-activedescendant="listbox-:r0:-option-0"
  className="smx-listbox border rounded"
  id="listbox-:r0:"
  onKeyDown={[Function]}
  role="listbox"
  tabIndex={0}
>
  <div
    aria-selected={true}
    className="px-3 py-1 cursor-pointer bg-primary-500 text-white"
    id="listbox-:r0:-option-0"
    onClick={[Function]}
    onMouseDown={[Function]}
    role="option"
  >
    One
  </div>
  <div
    aria-selected={false}
    className="px-3 py-1 cursor-pointer"
    id="listbox-:r0:-option-1"
    onClick={[Function]}
    onMouseDown={[Function]}
    role="option"
  >
    Two
  </div>
  <div
    aria-selected={false}
    className="px-3 py-1 cursor-pointer"
    id="listbox-:r0:-option-2"
    onClick={[Function]}
    onMouseDown={[Function]}
    role="option"
  >
    Three
  </div>
</div>
`;
